examples/casestudies/inference.effekt.md

```diff
@@ -0,0 +1,534 @@
+```
+def rejectionSampling[R](n: Int){program: () => R / {Sample, Observe}}: Unit = {
+  def loop(): Unit / Emit[R]= {
+    handleRandom{handleSample{handleRejection{handleObserve{
+      do emit(program())
+    }}}}
+    loop()
+  }
+  handleLimitEmit[R, Unit](n){loop()}
+}
+```
+
+## Slice Sampling
+```
+def sliceSamplingAlgo[R](){program: () => R / Weight} = {
+  val (result, prob) = handleRandom{handleSample{handleWeight{program()}}}
+  
+  def step(result0: R, prob0: Probability) = handleRejection{
+    val (result1, prob1) = handleWeight{program()}
+    if (prob1 < prob0) {do Weight(prob1 / prob0)}
+    (result1, prob1)
+  }
+
+  def loop(result: R, prob: Probability): Unit / Emit[R] = {
+    do emit(result)
+    val (result1, prob1) = step(result, prob)
+    loop(result1, prob1)
+  }
+
+  loop(result, prob)
+}
+```
+
+## Metropolis-Hastings
+### Tracing
+A trace records past samples used by an algorithm. The trace also controls where the algorithm draws samples from in the next iterations. 
+Constructing traces is possible by handling the effect `Sample`, not with the default handler, but one that draws new samples and records them in a trace.
+```
+def handleTracing[R]{program: () => R / Sample}: (R, Trace) / Sample = {
+  try {(program(), [])}
+  with Sample {(dist) => 
+    val d = do Sample(dist);
+    val (r, trace) = resume(d);
+    (r, Cons(d, trace))
+  }
+}
+
+def handleReusingTrace[R](trace0: Trace){program: () => R / Sample} = handleObserve{
+  var trace = trace0
+  try {program()}
+  with Sample {(dist) =>
+    trace match {
+      case Nil() => panic("empty trace")
+      case Cons(t, ts) => 
+        do Observe(t, dist)
+        trace = ts; resume(t) 
+    }
+  }
+}
+```
+### Proposing samples
+How the candidates of this algorithm are proposed can vary between different implementations of the algorithm and thus creating inference of the Metropolis-Hastings algorithm.
+The Metropolis-Hastings algorithm proposes a new trace based on the old trace, by recursively adding noise to the samples in the old trace and thus creating a new trace.
+```
+def propose(trace: Trace): Trace / Sample = {
+  trace match {
+    case Nil() => [] 
+    case Cons(t, ts) => 
+      val noise = do Sample(Gaussian(0.0, 1.0)) 
+      val proposal = t + noise
+      Cons(proposal, propose(ts)) 
+  }
+}
+```
+### Metropolis-Hastings algorithm
+The algorithm is implemented with a helper function
+```
+def metropolisStep[A](prob0: Probability, trace0: Trace){program: Trace => (A, Probability)} = {
+  val trace1 = propose(trace0)
+  val (result1, prob1) = program(trace1)
+
+  if(prob1 < prob0){
+    do Weight(prob1 / prob0)
+  }
+  ((result1, trace1), prob1)
+}
+```
+`metropolisStep` is then called in the algorithm implementation
+```
+def metropolisHastingsAlgo[A]{program: () => A / {Sample, Weight}} = {
+  val ((result0, trace0), prob0) = handleWeight{
+    handleRandom{
```

Comment:
Since `handleRandom` is implemented in terms of "global" randomness, the order of handlers here shouldn't matter, right? So we could remove `handleRandom` here and leave randomness uninterpreted.

Reply:
I think it would be possible to move `handleRandom` from this position to the wrapper `metropolisHastings`, but I'm not sure if it is possible to leave the randomness uninterpreted since `Random` is used to generate samples.
In terms of the order of the handlers, since `handleSample` (indirectly) uses the effect `Random`, `handleRandom` has to be applied after applying `handleSample`.

Reply:
I am proposing to move it completely to the outside, that is `main`. This way everything could in principle use randomness, but we can implement it in a deterministic manner.

Reply:
I am proposing to move it completely to the outside, that is `main`. This way everything could in principle use randomness, but we can implement it in a deterministic manner for testing.